### ISSUE TITLE
HTTPAuthResource: 'ValueError: too many values to unpack' when IRealm.requestAvatarId returns a logout callable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2522]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#2522|
|Type|defect|
|Created|2007-03-13 22:36:48Z|

IRealm.requestAvatar is supposed to return a tuple of (interface, avatar, logout) however HTTPAuthResource does ```interface, avatar = avatar``` after login and so will fail with a ValueError if the logout callable is returned.  HTTPAuthResource should allow a logout avatar to be returned (because it should always be returned according to the IRealm interface) and simply ignore it.  (There is no reasonable way to "logout" someone authenticated via HTTP Auth.)

Attachments:

* [wrapper_avatar_logout_fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/807/807c96dd078cc58d390e6eae0488d04429877634/ef1c662a47b8faf4c1c29a1a0cf08e5be64ff97d.patch) (446 bytes) - added by lann on 2008-06-12 13:59:46Z - Stupid patch that fixes this error

<details><summary>Searchable metadata</summary>

```
trac-id__2522 2522
type__defect defect
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__ 
time__1173825408000000 1173825408000000
changetime__1297611126000000 1297611126000000
version__None None
owner__dreid dreid
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

logout *should* be called; certain implementations will likely depend on it for resource cleanup. Perhaps it should be called on every query, or maybe have caching with timeout... but it should be called.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I would go so far as to say it must be called.  Perhaps the cred documentation should be clarified?  I think a reasonable expectation for an IRealm implementation to have is for logout to be called exactly once per requestAvatar call.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

```nevow.guard``` deals with this, fwiw.

**Comment by twisted-trac** [comment]
|<img alt="lann's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|lann commented|
|-|-|

The attached stupid patch resolves this error without breaking anything.

As for the logout semantics, I'm not sure that it makes sense in every situation, but I needed it for a long-running request (server push) that would call logout on connection close. I implemented it with a custom stream that had a callback on stream close(), but that might not make sense for some situations (e.g. pipelining).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

This code has recently been ported to `twisted.web` as part of [#10448](https://github.com/twisted/twisted/issues/10448), with correct handling of the `logout` callback.  It should be in trunk soon.  You may want to try that instead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

Not going to fix this in web2 (see [#4821](https://github.com/twisted/twisted/issues/4821)).  Works correctly in web, so closing this ticket.
